### PR TITLE
[Flow] Move first part of Flow transforms to new pipeline

### LINKED ISSUE #18099: Move dispatch region formation & preprocessing out of Flow
>2. Create a new "compilation phase" named CreateDispatchesDefault(iree-create-dispatches-default) just like Global optimizations/Flow/Stream/HAL, etc. that is essentially the transformations in the addDispatchRegionCreationPasses. So move those passes out of Flow into that phase. There is also an additional advantage of being able compile-from/compile-to this phase to isolate whats happening there in more complex models.

From the comments of https://github.com/iree-org/iree/pull/18063

There are a few inflight changes that need to land before this takes place

- https://github.com/iree-org/iree/pull/18063
- https://github.com/iree-org/iree/pull/18069
- https://github.com/iree-org/iree/pull/17880
- ...more?

cc @benvanik @MaheshRavishankar

### PR DESCRIPTION
**Pull request by IanWood1**
Refactor the logic in `Flow` that forms dispatches into a new `DispatchCreation` pipeline.

### CODE REVIEW

File: compiler/src/iree/compiler/Dialect/Flow/TransformExtensions/FlowExtensions.cpp [hunk 1]
```diff
@@ -8,8 +8,8 @@
 
 #include "iree/compiler/Dialect/Flow/IR/FlowOps.h"
 #include "iree/compiler/Dialect/Flow/Transforms/ConvertRegionToWorkgroups.h"
-#include "iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.h"
 #include "iree/compiler/Dialect/Flow/Transforms/RegionOpUtils.h"
+#include "iree/compiler/DispatchCreation/FormDispatchRegions.h"
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Dialect/Flow/TransformExtensions/FlowExtensions.cpp]: We probably dont want this dependence... Why did this need to be here?

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.cpp [hunk 2]
```diff
@@ -175,120 +114,10 @@ static void addCleanupPatterns(OpPassManager &passManager) {
...
-      .addPredicatedPass(clEnableDataTiling,
-                         [&]() {
-                           return createSetEncodingPass(
-                               SetEncodingPassOptions{clPadFactor});
-                         })
-      // Collapse dimensions of linalg Ops.
-      .addPass(IREE::Flow::createCollapseDimensionsPass)
-      // Convert dispatch regions into dispatch workgroups by capturing values
-      // and making the new workgroups isolated from above.
       .addPass(IREE::Flow::createConvertDispatchRegionsToWorkgroupsPass)
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.cpp]: Might be easier to move all the remaining passes also into dispatch region formation pipeline that you are adding. They are really specific to how we create dispatches by default. THat might help with the dependencies. 
You could still keep the utility to convert the `flow.dispatch.region` -> `flow.dispatch.workgroups` in Flow, but move the passes if that works.

File: compiler/src/iree/compiler/Preprocessing/Passes.cpp [hunk 3]
```diff
@@ -103,7 +104,7 @@ buildTransposeConvolutionPassPipeline(OpPassManager &passManager,
       .addPass(mlir::createLinalgNamedOpConversionPass)
       .addPass(GlobalOptimization::createConvert1X1FilterConv2DToMatmulPass)
       .addPass(createConvertConvToChannelsLastPass);
-  passManager.addPass(IREE::Flow::createFoldUnitExtentDimsPass());
+  passManager.addPass(DispatchCreation::createFoldUnitExtentDimsPass());
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Preprocessing/Passes.cpp:107]: Ugghh... this is making more clear wonky pass dependencies we have

**Comment by IanWood1** [comment]
Should the dispatch_linalg_on_tensors* tests be using the new pipeline? 

Because https://github.com/iree-org/iree/blob/292f2d400506acda3fa5edc76074d630eda66a98/compiler/src/iree/compiler/Dialect/Flow/Transforms/test/dispatch_linalg_on_tensors.mlir has a lot of specific 'CHECK' statements and the `iree-opt` pass pipeline doesn't exactly match the pipeline's passes. I wasn't able to get it to work

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp [hunk 4]
```diff
@@ -80,9 +53,6 @@ void TensorDimTrackingRewriter::notifyOperationInserted(Operation *op,
 
 namespace mlir::iree_compiler::IREE::Flow {
 
-#define GEN_PASS_DEF_FORMDISPATCHREGIONSPASS
-#include "iree/compiler/Dialect/Flow/Transforms/Passes.h.inc"
-
 LogicalResult simplifyDimOps(RewriterBase &rewriter,
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp:56]: Just for future. There is an update for `simplifyDimOps` that we need to do to truly handle dynamic cases.. we should chat about it sometime.

File: compiler/src/iree/compiler/DispatchCreation/Passes.td [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+include "mlir/Pass/PassBase.td"
+
+//===---------------------------------------------------------------------===//
+// Dispatch region creation preprocessing passes :
+// Passes that transform the program before forming dispatches, like
+// - Elementwise operation fusion
+// - Reshape propagation passes
+//===---------------------------------------------------------------------===//
+
+def TensorPadToTensorInsertSlicePass :
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/DispatchCreation/Passes.td:20]: Could you move over the file organization comment from Flow and make sure these are "alphabetically ordered in each section". Also fine to do this as a follow up.

**Review by MaheshRavishankar — APPROVED** [review]
Ok. Looks good to me. Thanks for doing this! We might have some follow up fixes here, but its good to make the dispatch region creation its own "compilation phase".

**Comment by IanWood1** [comment]
Okay, I'll make those changes as a follow up. I'll create a new issue to track the upcoming follow-up changes.

**Pull request merged**